src/main/java/com/google/devtools/build/lib/actions/cache/CompactPersistentActionCache.java

```diff
@@ -423,66 +452,97 @@ public synchronized void dump(PrintStream out) {
       }
       out.println(entry.getKey() + ", " + indexer.getStringForIndex(entry.getKey()) + ":\n"
-          +  content + "\n      packed_len = " + entry.getValue().length + "\n");
+          + content + "\n      packed_len = " + entry.getValue().length + "\n");
     }
   }
 
   /**
    * @return action data encoded as a byte[] array.
    */
-  private static byte[] encode(StringIndexer indexer, ActionCache.Entry entry) {
+  private static byte[] encode(
+      StringIndexer indexer, ObjectCodecs objectCodecs, ActionCache.Entry entry) throws IOException {
     Preconditions.checkState(!entry.isCorrupted());
 
-    try {
-      byte[] actionKeyBytes = entry.getActionKey().getBytes(ISO_8859_1);
-      Collection<String> files = entry.getPaths();
-
-      // Estimate the size of the buffer:
-      //   5 bytes max for the actionKey length
-      // + the actionKey itself
-      // + 32 bytes for the digest
-      // + 5 bytes max for the file list length
-      // + 5 bytes max for each file id
-      // + 32 bytes for the environment digest
-      int maxSize =
-          VarInt.MAX_VARINT_SIZE
-              + actionKeyBytes.length
-              + DigestUtils.ESTIMATED_SIZE
-              + VarInt.MAX_VARINT_SIZE
-              + files.size() * VarInt.MAX_VARINT_SIZE
-              + DigestUtils.ESTIMATED_SIZE;
-      ByteArrayOutputStream sink = new ByteArrayOutputStream(maxSize);
-
-      VarInt.putVarInt(actionKeyBytes.length, sink);
-      sink.write(actionKeyBytes);
-
-      MetadataDigestUtils.write(entry.getFileDigest(), sink);
-
-      VarInt.putVarInt(entry.discoversInputs() ? files.size() : NO_INPUT_DISCOVERY_COUNT, sink);
-      for (String file : files) {
-        VarInt.putVarInt(indexer.getOrCreateIndex(file), sink);
+    byte[] actionKeyBytes = entry.getActionKey().getBytes(ISO_8859_1);
+    Collection<String> files = entry.getPaths();
+
+    // Estimate the size of the buffer:
+    //   5 bytes max for the actionKey length
+    // + the actionKey itself
+    // + 32 bytes for the digest
+    // + 5 bytes max for the file list length
+    // + 5 bytes max for each file id
+    // + 32 bytes for the environment digest
+    int maxSize =
+        VarInt.MAX_VARINT_SIZE
+            + actionKeyBytes.length
+            + DigestUtils.ESTIMATED_SIZE
+            + VarInt.MAX_VARINT_SIZE
+            + files.size() * VarInt.MAX_VARINT_SIZE
+            + DigestUtils.ESTIMATED_SIZE;
+    ByteArrayOutputStream sink = new ByteArrayOutputStream(maxSize);
+
+    VarInt.putVarInt(actionKeyBytes.length, sink);
+    sink.write(actionKeyBytes);
+
+    MetadataDigestUtils.write(entry.getFileDigest(), sink);
+
+    VarInt.putVarInt(entry.discoversInputs() ? files.size() : NO_INPUT_DISCOVERY_COUNT, sink);
+    for (String file : files) {
+      VarInt.putVarInt(indexer.getOrCreateIndex(file), sink);
+    }
+
+    MetadataDigestUtils.write(entry.getUsedClientEnvDigest(), sink);
+
+    VarInt.putVarInt(entry.getOutputFiles().size(), sink);
```

Comment:
I like that we always store the empty maps. You don't need to make a distinction on the cache format thanks to that. Is there a noticeable difference in the size of the cache with the flag disable due to the extra 2 ints?

Reply:
Nope. For building Bazel itself, the size of action cache increased from 896K to 904K (0.9%). When the store flag is set, the size increased to 1.8M.